crates/sui-storage/src/indexes.rs

```diff
@@ -334,22 +412,44 @@ impl IndexStore {
                     obj_id, written_coins, digest
                 )
             });
-            let coin = obj.as_coin_maybe().unwrap_or_else(|| {
+            let coin_balance = obj.move_object().and_then(|mo| mo.get_coin_value().ok())
+            .unwrap_or_else(|| {
                 panic!(
-                    "object_id: {:?} in written_coins cannot be deserialzied as a Coin, written_coins: {:?}, tx_digest: {:?}",
+                    "object_id: {:?} in written_coins is not a coin type, written_coins: {:?}, tx_digest: {:?}",
                     obj_id, written_coins, digest
                 )
             });
-            address_coin_type_to_invalidate.insert((*owner, coin_type_tag.clone()));
+            let coin_type_str = coin_type_tag.to_string();
+            address_coin_type_to_invalidate.insert((*owner, coin_type_str.clone()));
             addresses_to_invalidate.insert(*owner);
-            Some(((*owner, coin_type_tag.to_string(), *obj_id), (CoinInfo {version: obj_info.version, digest: obj_info.digest, balance: coin.balance.value(), previous_transaction: *digest})))
+            Some(((*owner, coin_type_str, *obj_id), (CoinInfo {version: obj_info.version, digest: obj_info.digest, balance: coin_balance, previous_transaction: *digest})))
         }).collect::<Vec<_>>();
+
         trace!(
             tx_digset=?digest,
             "coin_add_keys: {:?}",
             coin_add_keys,
         );
-        batch.insert_batch(&self.tables.coin_index, coin_add_keys.into_iter())?;
+        batch.insert_batch(&self.tables.coin_index, coin_add_keys.clone().into_iter())?;
+        batch.merge_batch(
+            &self.tables.balance_index,
+            coin_add_keys
+                .into_iter()
+                .map(|((addr, coin_type, obj_id), coin_info)| {
+                    // If the balance is not found in input, it means the object does not belong to an AddressOwner
+                    // prior to this transasction. In this case, we add the full balance of the coin,
+                    // and 1 additional coin. Otherwise, we only add the delta (could be negative), and
+                    // keep the coin count intact.
+                    let input_balance = input_coin_balances.get(&obj_id);
+                    let balance_delta = coin_info.balance as i128
+                        - (**input_balance.as_ref().unwrap_or(&&0u64) as i128);
+                    (
+                        (addr, coin_type),
+                        (balance_delta, 1i64 - input_balance.map_or(0i64, |_| 1i64)),
+                    )
```

Comment:
we need to check the ownership change here. If the coin is transferred to a new owner, then the owner can just add full amount of the balance.